### PR TITLE
(PUP-8008) Avoid double-encoding Forge API pagination URIs

### PR DESCRIPTION
**Pull request by scotje**
This should resolve the issue with PMT following the pagination links returned by the Forge API.

### CODE REVIEW

**Comment by scotje** [comment]
Oops, looks like there are 2 sets of unit tests for the Forge code. I'll fix up the other tests.

File: lib/puppet/forge.rb [hunk 1]
```diff
@@ -90,9 +90,9 @@ def search(term)
   # @see SemanticPuppet::Dependency::Source#fetch
   def fetch(input)
     name = input.tr('/', '-')
-    uri = "/v3/releases?module=#{name}"
+    uri = "/v3/releases?module=#{name}&sort_by=version"
```
Comment (MosesMendoza) [inline lib/puppet/forge.rb]: Out of curiosity, why did you add the sort by query param here?

File: lib/puppet/forge.rb [hunk 2]
```diff
@@ -90,9 +90,9 @@ def search(term)
...
     if Puppet[:module_groups]
-      uri += "&module_groups=#{Puppet[:module_groups].gsub('+', ' ')}"
+      uri += "&module_groups=#{Puppet[:module_groups]}"
```
Comment (MosesMendoza) [inline lib/puppet/forge.rb]: is there any downside to leaving these gsub's here in case something else is using this uri?

File: lib/puppet/forge.rb (hunk 1 above)
Comment (scotje) [reply lib/puppet/forge.rb]: I'll add some info to the commit message.

File: lib/puppet/forge.rb (hunk 2 above)
Comment (scotje) [reply lib/puppet/forge.rb]: Probably not?

**Comment by scotje** [comment]
Updated commit message for `sort_by` change.

File: lib/puppet/forge.rb (hunk 2 above)
Comment (scotje) [reply lib/puppet/forge.rb]: Re-added the `gsub`s.

File: lib/puppet/forge.rb (hunk 1 above)
Comment (Iristyle) [reply lib/puppet/forge.rb]: Can you query directly by version number here?

Comment (scotje) [reply lib/puppet/forge.rb]: Yes but PMT currently collects all the available versions in case it needs them when resolving dependencies.

File: lib/puppet/forge.rb [hunk 3]
```diff
@@ -113,8 +113,14 @@ def fetch(input)
     return releases
   end
 
-  def make_http_request(*args)
-    @repository.make_http_request(*args)
+  def make_http_request(uri, *args)
+    # Change plus to space so it can be encoded properly
+    uri = uri.gsub('+', ' ')
```
Comment (Iristyle) [inline lib/puppet/forge.rb]: Since this is called by `download` as well, I'm not sure that this is the safest location to make this change.  I think it was intentional to make that mod inline before.

**Comment by Iristyle** [comment]
I think it would be good to look at a couple of things before making changes here:

* Can we upload a module to the forge that contains UTF-8 characters (and will match a local query) for the sake of some additional testing / debugging?
* Can we look at how the old PMT client behaved ... prior to any of these changes went in? Something still doesn't totally add up here for me, and I think it would helpful to have the actual client / server interactions (at least in terms of request URIs) to understand this better.
    * For instance, in Puppet 4.8,  the request trickles through to https://github.com/puppetlabs/puppet/blob/3deda986fcbdf62fe9c3880d3a6c7b91b5122ed3/lib/puppet/forge/repository.rb#L76 - `request = Net::HTTP::Get.new(Puppet::Util.uri_encode(path), headers)`
    * In current code, the request trickles through to https://github.com/puppetlabs/puppet/blob/master/lib/puppet/forge/repository.rb#L76 - `request = Net::HTTP::Get.new(Puppet::Util.uri_encode(path), headers)`
    * Those basically look the same, and we weren't decoding URIs before... what exactly about the interaction has changed? (Is it just the bit about the `+` or were already things subtly wrong before that we just didn't run into)
* What other clients are consuming the JSON payload? Given the request code mentioned in prior bullet, why do URIs need to be passed encoded in the JSON, given they're being encoded again on the way out of PMT?

**Comment by scotje** [comment]
> Can we upload a module to the forge that contains UTF-8 characters (and will match a local query) for the sake of some additional testing / debugging?

In the actual module name? I don't think non-ASCII characters have ever been allowed in module names on the Forge. (See https://github.com/puppetlabs/puppet-forge-api/blob/master/app/validators.rb for the accepted regexes) But we can upload whatever we want to one of the non-public Forge instances for testing.

> Can we look at how the old PMT client behaved ... prior to any of these changes going in? Something still doesn't totally add up here for me, and I think it would helpful to have the actual client / server interactions (at least in terms of request URIs) to understand this better. For instance, in Puppet 4.8...

I'm guessing the pagination has probably been broken since then? It only manifests if the requested release isn't in the first page of results so it may have taken that long for the issue to surface. 

The previous commit to that:
https://github.com/puppetlabs/puppet/blob/5ce0917cf8b63dc682987aab9e93976509dd284f/lib/puppet/forge/repository.rb#L75

Uses `URI.escape` which leaves the pluses alone:

```
irb(main):002:0> URI.escape('module_groups=base+pe_only')
=> "module_groups=base+pe_only"
```

You can see the issue pretty clearly if you adjust the debug logging the way I did in the PR without any of the other changes:

```
$ bx bin/puppet module install puppetlabs-apache -v 2.3.0 --modulepath=/Users/jesse/sandbox/pmt --debug --trace --module_groups="base+pe_only"
Debug: Runtime environment: puppet_version=5.3.2, ruby_version=2.4.0, run_mode=user, default_encoding=UTF-8
Notice: Preparing to install into /Users/jesse/sandbox/pmt ...
Notice: Downloading from https://forgeapi.puppet.com ...
Debug: Evicting cache entry for environment 'production'
Debug: Caching environment 'production' (ttl = 0 sec)
Debug: HTTP GET https://forgeapi.puppet.com/v3/releases?module=puppetlabs-apache&module_groups=base%20pe_only
Debug: HTTP GET https://forgeapi.puppet.com/v3/releases?module=puppetlabs-apache&module_groups=base%2Bpe_only&limit=20&offset=20
Debug: HTTP GET https://forgeapi.puppet.com/v3/releases?module=puppetlabs-stdlib&module_groups=base%20pe_only
Debug: HTTP GET https://forgeapi.puppet.com/v3/releases?module=puppetlabs-stdlib&module_groups=base%2Bpe_only&limit=20&offset=20
Debug: HTTP GET https://forgeapi.puppet.com/v3/releases?module=puppetlabs-concat&module_groups=base%20pe_only
Debug: HTTP GET https://forgeapi.puppet.com/v3/releases?module=puppetlabs-concat&module_groups=base%2Bpe_only&limit=20&offset=20
Debug: HTTP GET https://forgeapi.puppet.com/v3/releases?module=puppetlabs-firewall&module_groups=base%20pe_only
Debug: HTTP GET https://forgeapi.puppet.com/v3/releases?module=puppetlabs-firewall&module_groups=base%2Bpe_only&limit=20&offset=20
Error: Could not install 'puppetlabs-apache' (v2.3.0)
  No releases matching '2.3.0' are available from https://forgeapi.puppet.com
```

Specifically, the first request for each page has `module_groups=base%20pe_only` (the plus was converted to a space and then encoded) but the subsequent requests have `module_groups=base%2Bpe_only` (literal plus encoded as `%2B`).

> What other clients are consuming the JSON payload? Given the request code mentioned in prior bullet, why do URIs need to be passed encoded in the JSON, given they're being encoded again on the way out of PMT?

It's a public API that has been documented and recommended for ~4 years so it's hard to definitively answer that question. At the least there is `r10k`, `forge-ruby`, `librarian-puppet`, the new Artifactory stuff.

**Comment by scotje** [comment]
Closing this, opening a new fix targeted at 4.10.x first.
